src/Badges.t.sol

```diff
@@ -0,0 +1,521 @@
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.6;
+
+import "forge-std/Test.sol";
+import { IERC165 } from "./IERC165.sol";
+
+import { IERC721Metadata } from "./IERC721Metadata.sol";
+import { IERC4973 } from "ERC4973/interfaces/IERC4973.sol";
+import { Badges } from "./Badges.sol";
+import { SpecDataHolder } from "./SpecDataHolder.sol";
+import { Raft } from "./Raft.sol";
+
+contract ERC1271Mock {
+  bytes4 internal constant MAGICVALUE = 0x1626ba7e;
+  bool private pass;
+
+  constructor(bool pass_) {
+    pass = pass_;
+  }
+
+  function isValidSignature(bytes32 hash, bytes memory signature) public view returns (bytes4) {
+    if (pass) {
+      return MAGICVALUE;
+    } else {
+      revert("permit not granted");
+    }
+  }
+}
+
+contract AccountAbstraction is ERC1271Mock {
+  constructor(bool pass) ERC1271Mock(pass) {}
+
+  function give(
+    address collection,
+    address to,
+    string calldata uri,
+    bytes calldata signature
+  ) external virtual returns (uint256) {
+    return Badges(collection).give(to, uri, signature);
+  }
+
+  function take(
+    address collection,
+    address from,
+    string calldata uri,
+    bytes calldata signature
+  ) external virtual returns (uint256) {
+    return Badges(collection).take(from, uri, signature);
+  }
+
+  function unequip(address collection, uint256 tokenId) public virtual {
+    return Badges(collection).unequip(tokenId);
+  }
+}
+
+contract NonAuthorizedCaller {
+  function unequip(address collection, uint256 tokenId) external {
+    Badges badges = Badges(collection);
+    badges.unequip(tokenId);
+  }
+}
+
+contract BadgesTest is Test {
+  ERC1271Mock approver;
+  ERC1271Mock rejecter;
+  Badges badges;
+  SpecDataHolder specDataHolder;
+  Raft raft;
+  AccountAbstraction aa;
+
+  address passiveAddress = 0x0f6A79A579658E401E0B81c6dde1F2cd51d97176;
+  uint256 passivePrivateKey = 0xad54bdeade5537fb0a553190159783e45d02d316a992db05cbed606d3ca36b39;
+  uint256 randomPrivateKey = 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;
+  string specUri;
+
+  event Transfer(address indexed from, address indexed to, uint256 indexed tokenId);
+
+  function setUp() public {
+    address to = address(this);
+
+    badges = new Badges();
+    specDataHolder = new SpecDataHolder();
+    raft = new Raft();
+    badges.initialize("Badges", "BADGES", "0.1.0", to, address(specDataHolder));
+    raft.initialize(to, "Raft", "RAFT");
+    specDataHolder.initialize(address(raft), to);
+
+    approver = new ERC1271Mock(true);
+    rejecter = new ERC1271Mock(false);
+    aa = new AccountAbstraction(true);
+
+    specUri = "some spec uri";
+
+    vm.label(passiveAddress, "passive");
+  }
+
+  // helper function
+  function createRaftAndRegisterSpec() internal {
```

Comment:
Move internal funcs to after public funcs per styleguide
https://docs.soliditylang.org/en/latest/style-guide.html#order-of-functions